app/models/service/aggregation.rb

```diff
@@ -9,12 +9,243 @@ module Service::Aggregation
+    # @option (see #vm_join_clause)
+    # @return [Arel::SelectManager] Arel with joins statements for
+    #   service_resources, vms, and hardwares
+    def base_service_aggregation_join(arel, services_tbl, options = {})
+      arel.join(service_resources_tbl).on(service_resources_tbl[:service_id].eq(services_tbl[:id])
+                                  .and(service_resources_tbl[:resource_type].eq(vm_or_template_type)))
+          .join(vms_tbl).on(vm_join_clause(options))
+          .join(hardwares_tbl).on(hardwares_tbl[:vm_or_template_id].eq(vms_tbl[:id]))
+    end
+
+    # Generates the following SQL conditional to be used in
+    # `base_service_aggrigation_join`:
+    #
+    #     "vms"."id" = "service_resources"."resource_id"
+    #
+    # But will append an "AND" if `on_vms_only` is passed in as a option:
+    #
+    #     "vms"."id" = "service_resources"."resource_id" AND LOWER("vms"."power_state") == "on"
+    #
+    #
+    # @option options [Boolean] :on_vms_only Filter "on" VMs only in the JOIN
+    # @return [Arel::Nodes::And] The Arel conditional for JOINing the VMs table
+    def vm_join_clause(options = {})
+      clause = vms_tbl[:id].eq(service_resources_tbl[:resource_id])
+      if options[:on_vms_only] # meaning VMs that are powered "on"
+        clause = clause.and(vms_tbl[:power_state].lower.eq('on'))
+      end
+      clause
+    end
+
+    # Generates the following JOIN statement in arel
+    #
+    #    JOIN disks ON disks.hardware_id = hardwares.id
+    #
+    # @param (see #base_service_aggrigation_join)
+    # @return (see #base_service_aggrigation_join)
+    def join_on_disks(arel)
+      arel.join(disks_tbl).on(disks_tbl[:hardware_id].eq(hardwares_tbl[:id]))
+    end
+
+    # NOTE: The following class methods are technically "constants", but to
+    # make sure they are not triggered when the class is first loaded and cause
+    # requests to the DB (can break building the appliance), we are making them
+    # memoized class methods.
+
+    def service_resources_tbl
+      @service_resources_tbl ||= ServiceResource.arel_table
+    end
+
+    def vms_tbl
+      @vms_tbl ||= Vm.arel_table
+    end
+
+    def hardwares_tbl
+      @hardwares_tbl ||= Hardware.arel_table
+    end
+
+    def disks_tbl
+      @disks_tbl ||= Disk.arel_table
+    end
+
+    def partitions_tbl
+      @partitions_tbl ||= Partition.arel_table
+    end
+
+    def zero
+      @zero ||= Arel::Nodes::SqlLiteral.new("0")
+    end
+
+    def vm_or_template_type
+      @vm_or_template_type ||= Arel::Nodes::SqlLiteral.new("'VmOrTemplate'")
+    end
+
+    def ancestry_match
+      @ancestry_match ||= Arel::Nodes::SqlLiteral.new("'/%'")
+    end
+
+    def ancestry_ilike
+      @ancestry_ilike ||= Arel::Nodes::NamedFunction.new("CONCAT", [Service.arel_table[:id], ancestry_match])
+    end
+
+    def service_id_cast
+      @service_id_cast ||= Arel::Nodes::NamedFunction.new("CAST", [Service.arel_table[:id].as("VARCHAR")])
```

Comment:
@Fryguy yes, you are correct since we are comparing against the ancestry column, since it is a string:

https://github.com/ManageIQ/manageiq/blob/master/db/migrate/20130923182042_collapsed_initial_migration.rb#L2583

https://github.com/NickLaMuro/manageiq/blob/99ff33aa/app/models/service/aggregation.rb#L153